### ISSUE TITLE
Empty `gr.Dataframe` is not empty

### ISSUE THREAD

**Issue by hysts** [gradio-app/gradio#5162]
### Describe the bug

When passing an empty dataframe, say `pd.DataFrame([], columns=['a', 'b'])`, to `gr.Dataframe`, the value becomes non-empty dataframe `pd.DataFrame(['', ''], columns=['a', 'b'])`.

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

```python
import gradio as gr
import pandas as pd

df_orig = pd.DataFrame([], columns=['a', 'b'])

with gr.Blocks() as demo:
    df = gr.Dataframe(value=df_orig, interactive=False)
    btn = gr.Button()
    text = gr.Textbox()
    btn.click(fn=lambda x: x.shape, inputs=df, outputs=text)
demo.queue().launch()
```
https://huggingface.co/spaces/hysts-debug/empty-dataframe

An empty row is shown in the table view, and it says the shape is `(1, 2)` when clicking the button, but the table shouldn't show the empty row and the shape should be `(0, 2)`.

### Screenshot

![](https://github.com/gradio-app/gradio/assets/25161192/386696b6-c0c0-4433-b893-d91408bf1b5a)

### System Info

```shell
gradio==3.39.0
```

### Severity

I can work around it

**Comment by abidlabs** [comment]
When trying to reproduce this, I noticed that I can't run your example fully. Clicking on the button gives me:

> ValueError: 3 columns passed, passed data had 2 columns

Can you retry your example on the latest version of `gradio`?

**Comment by hysts** [comment]
Thanks. Yeah, looks like the code raises different error when using `gradio==3.40.1`. https://huggingface.co/spaces/hysts-debug/empty-dataframe-gradio-3-40-1
(FYI, the reproduction with the old `gradio==3.39.0` is here: https://huggingface.co/spaces/hysts-debug/empty-dataframe)

The error I get with `gradio==3.40.1` is
```
Running on local URL:  http://0.0.0.0:7860

To create a public link, set `share=True` in `launch()`.
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/pandas/core/internals/construction.py", line 934, in _finalize_columns_and_data
    columns = _validate_or_indexify_columns(contents, columns)
  File "/home/user/.local/lib/python3.10/site-packages/pandas/core/internals/construction.py", line 981, in _validate_or_indexify_columns
    raise AssertionError(
AssertionError: 3 columns passed, passed data had 2 columns

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/gradio/routes.py", line 488, in run_predict
    output = await app.get_blocks().process_api(
  File "/home/user/.local/lib/python3.10/site-packages/gradio/blocks.py", line 1428, in process_api
    inputs = self.preprocess_data(fn_index, inputs, state)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/blocks.py", line 1245, in preprocess_data
    processed_input.append(block.preprocess(inputs[i]))
  File "/home/user/.local/lib/python3.10/site-packages/gradio/components/dataframe.py", line 199, in preprocess
    return pd.DataFrame(x["data"], columns=x.get("headers"))
  File "/home/user/.local/lib/python3.10/site-packages/pandas/core/frame.py", line 782, in __init__
    arrays, columns, index = nested_data_to_arrays(
  File "/home/user/.local/lib/python3.10/site-packages/pandas/core/internals/construction.py", line 498, in nested_data_to_arrays
    arrays, columns = to_arrays(data, columns, dtype=dtype)
  File "/home/user/.local/lib/python3.10/site-packages/pandas/core/internals/construction.py", line 840, in to_arrays
    content, columns = _finalize_columns_and_data(arr, columns, dtype)
  File "/home/user/.local/lib/python3.10/site-packages/pandas/core/internals/construction.py", line 937, in _finalize_columns_and_data
    raise ValueError(err) from err
ValueError: 3 columns passed, passed data had 2 columns
```
and the screenshot of the demo is
![](https://github.com/gradio-app/gradio/assets/25161192/b4659b69-f166-42df-88ef-066aa587d3d2)
which is very weird because the table has somehow three columns now, though the passed dataframe only has two columns.

If it would be better to close this issue and open another one for this new error, feel free to do so.

**Comment by abidlabs** [comment]
Ok I agree its weird, I'll take a look!

**Comment by hysts** [comment]
Looks like the new error is due to https://github.com/gradio-app/gradio/issues/5237, and the original issue still exists with `gradio==3.40.1`.

reproduction: https://huggingface.co/spaces/hysts-debug/empty-dataframe-2

```python
import gradio as gr
import pandas as pd

df_empty = pd.DataFrame([], columns=['a', 'b'])

with gr.Blocks() as demo:
    df = gr.Dataframe(interactive=False)
    btn = gr.Button()
    text = gr.Textbox()
    btn.click(fn=lambda x: (gr.update(value=df_empty), len(x)),
              inputs=df,
              outputs=[df, text])
demo.queue().launch()
```

By clicking the button, the dataframe is updated to empty one, but the table shows an empty row and the number of rows is 1.
![](https://github.com/gradio-app/gradio/assets/25161192/8004b9fb-8550-4b0a-80a6-a4bc804920f8)

**Comment by abidlabs** [comment]
Taking a look!

**Comment by abidlabs** [comment]
@hysts so this is an interesting issue. There is indeed a bug, which will be fixed by #5256. But you also do need to change your code a bit and add `row_count = (0, "dynamic")` to your `gr.DataFrame`. The reason is that the `gr.DataFrame` by default sets a minimum row count to 1, so you would see an extra row being created. 

Here's the fixed code (will work on `main` after we merge in the PR):

```py
import gradio as gr
import pandas as pd

df_orig = pd.DataFrame([], columns=['a', 'b'])

with gr.Blocks() as demo:
    df = gr.Dataframe(value=df_orig, interactive=False, row_count=(0, "dynamic"))
    btn = gr.Button()
    text = gr.Textbox()
    btn.click(fn=lambda x: x.shape, inputs=df, outputs=text)
demo.queue().launch()
```

**Comment by hysts** [comment]
Oh, I didn't know that. Thanks a lot!
